### PR TITLE
Second group of module config sections

### PR DESCRIPTION
**Pull request by abea**

### CODE REVIEW

File: docs/reference/module-api/module-overview.md [hunk 1]
```diff
@@ -257,14 +257,231 @@ modules.export = {
 };
 ```
 
-## Customization functions
+## Initialization function
 
 ### `async init(self)`
-### `beforeSuperClass(self)`
+
+This function runs once when the Apostrophe app first start up. It takes the module, as `self`, as an argument.
```
Comment (boutell) [inline docs/reference/module-api/module-overview.md:264]: first starts up.

Comment (boutell) [reply docs/reference/module-api/module-overview.md:264]: Based on years and years of developer emails and questions, I think it's worth adding "it does NOT run on every request. For that, see event handlers, specifically `@apostrophecms/page:beforeSend`."

Comment (boutell) [reply docs/reference/module-api/module-overview.md:264]: Many new Apostrophe developers don't really get that node servers are long-lasting processes.

File: docs/reference/module-api/module-overview.md [hunk 2]
```diff
@@ -257,14 +257,231 @@ modules.export = {
...
+```javascript
+// modules/product/index.js
+module.exports = {
+  // ...
+  options: {
+    alias: 'product',
+    themeColor: 'blue'
+  },
+  init(self) {
+    // 👇 Making `self.options.themeColor` available on `apos.product.theme`.
```
Comment (boutell) [inline docs/reference/module-api/module-overview.md:277]: Nice.

File: docs/reference/module-api/module-overview.md [hunk 3]
```diff
@@ -257,14 +257,231 @@ modules.export = {
...
+    self.theme = self.options.themeColor;
+
+    // 👇 Adding a data migration related to this module using a method from a
+    // separate module.
+    self.apos.migration.add('blurb', async () => {
+      await self.apos.migration.eachDoc({
+        type: 'product'
+      }, async (doc) => {
+        if ((doc.category) === 'old-category') {
+          doc.category = 'new-category';
```
Comment (boutell) [inline docs/reference/module-api/module-overview.md:287]: Setting doc.category in this way has no effect on the database, only what is in updateOne matters, so this should become part of the `$set` object.

File: docs/reference/module-api/module-overview.md [hunk 4]
```diff
@@ -257,14 +257,231 @@ modules.export = {
...
+### The extension pattern
+
+Several of these sections use an extention pattern (the sections prefixed with "extend"). These sections are used to add functionality to matching features of the module's base class. Functions in `extendMethods` enhance identically named functions in the base class' `methods` section, `extendApiRoutes` enhances API routes from the base class' `apiRoutes`, and so on.
+
+Each individual function included in the extension section takes a `_super` argument in addition to the same arguments as the original function. `_super` is the original function, which should be called within the new extension.
+
+If a piece type included the `insert` in its `extendMethods` section to alter the piece titles, it might look like this:
+
+```javascript
+  enhanceMethods(self) {
```
Comment (boutell) [inline docs/reference/module-api/module-overview.md:336]: `extendMethods`, not `enhanceMethods`.

**Review by boutell — CHANGES_REQUESTED** [review]
Great stuff, minor things.

File: docs/reference/module-api/module-overview.md (hunk 1 above)
Comment (abea) [reply docs/reference/module-api/module-overview.md:264]: Good point. This isn't specifically about pages, so I don't think we should point to that specifically, but we can link over to an event handlers guide.

Comment (abea) [reply docs/reference/module-api/module-overview.md:264]: Adding "To run code on every request, or in other situations when the app is running, see the event handlers documentation." Let me know if there's other language for that you think would be more appropriate.

File: docs/reference/module-api/module-overview.md (hunk 3 above)
Comment (abea) [reply docs/reference/module-api/module-overview.md:287]: Yep. Lazy code copying.

File: docs/reference/module-api/module-overview.md [hunk 5]
```diff
@@ -257,14 +257,231 @@ modules.export = {
...
+        piece.title = `🆕 ${piece.title}`;
+
+        _super(req, piece, options);
+      }
+    }
+  }
+```
+
+::: warning
+Extension functions should *always* use the `_super` argument to call the original function. If you want to *completely* overwrite the inherited function, instead add a matching function in the section without the "extend" prefix. For example, we could completely overwrite the `insert` method in our piece type by including our own `insert` function in the `methods` section.
```
Comment (localghost443) [inline docs/reference/module-api/module-overview.md:348]: "instead add a matching function in the section without the "extend" prefix" -> "add a matching function in the section without the "extend" prefix instead".

File: docs/reference/module-api/module-overview.md [hunk 6]
```diff
@@ -257,14 +257,231 @@ modules.export = {
...
+        return piece;
+      }
+    };
+  }
+};
+```
+
 ### `components(self)`
+
+This function section returns an object containing functions that power asynchronous template components. These template components allow for asynchronous data requests in normally synchronous template rendering.
```
Comment (localghost443) [inline docs/reference/module-api/module-overview.md:411]: "This function section" doesn't seem necessary. "These template components" -> "asynchronous template component".

**Review by boutell — APPROVED** [review]

**Review by localghost443 — APPROVED** [review]

**Pull request merged**